src/bit.c

```diff
@@ -159,16 +160,18 @@ int bit_load_firmware(struct bit* bit, struct cha* cha, struct chb* chb) {
 	while (length) {
 		size_t i = 0;
 
-		for (i = 0; i < sizeof(buf) && length; i++, length--, data++) {
-			buf[i] = bitreverse8(*data);
+		bidx = (bidx + 1) % 2;
+
+		for (i = 0; i < sizeof(buf[bidx]) && length; i++, length--, data++) {
+			buf[bidx][i] = bitreverse8(*data);
 		}
 
 		if (tc && ftdi_transfer_data_done(tc) < 0) {
 			bit->error_str = ftdi_get_error_string(&cha->ftdi);
 			return -1;
 		}
 
-		if (!(tc = ftdi_write_data_submit(&cha->ftdi, buf, i))) {
+		if (!(tc = ftdi_write_data_submit(&cha->ftdi, buf[bidx], i))) {
```

Comment:
I am afraid but it is so low level that I am not aware of any single doc line explaining the reason. Thus, we have to dig deep into the code:
- ftdi_write_data_submit() calls libusb_submit_transfer() under the hood
- libusb_submit_transfer() calls one of its backend function. On Windows the execution eventually ends up in winusbx_submit_bulk_transfer()
- winusbx_submit_bulk_transfer() uses WritePipe with the OVERLAPPED structure
- control returns

Notice that the data was never copied and it is the original buffer that's submitted. Thus the driver will write to the provided buffer. On Windows, kernel code executes at PASSIVE_LEVEL or higher. Driver code operating at PASSIVE_LEVEL can basically interrupt any user application at any time. What happens here, is the driver code writes to the buffer before we read it.

Thus we have to either use two buffers (as done here), or simply use synchronous ftdi_write_data().

I do not consider this to be a Windows only issue. This basically can happen in any system that does not copy the buffer.

Reply:
Hm... Now it is getting clear to me. Thank you. I get used that in Linux the data is copied to the kernel at the syscall in 99% cases.

Reply:
Can we rephrase the patch as the following: `int bidx` -> `size_t bidx` (since bidx is used as array index variable)?
And redo `while(length) {` loop as the following
```
for (bidx = 0; length; bidx = (bidx + 1) % 2) {
```

Reply:
I have applied the `size_t` and `while` -> `for` change mentioned in above comment.